File: patcha/process.py
"""Event preprocessing — text extraction and embedding generation."""

import json
import logging
from pathlib import Path
from typing import List, Optional

from patcha import embedding
from patcha.config import config
from patcha.db.models import Event, EventType
from patcha.utils.chunking import chunk_text

log = logging.getLogger(__name__)

_PENDING_PATH = config.data_dir / "pending_events.jsonl"


class PendingEventStore:
    def __init__(self, path: Path = _PENDING_PATH) -> None:
        self._path = path
        self._path.parent.mkdir(parents=True, exist_ok=True)

    def save(self, event: Event) -> None:
        with open(self._path, "a") as f:
            f.write(event.model_dump_json() + "\n")

    def load_all(self) -> List[Event]:
        if not self._path.exists():
            return []
        events = []
        for line in self._path.read_text().splitlines():
            line = line.strip()
            if line:
                try:
                    events.append(Event.model_validate_json(line))
                except Exception as e:
                    log.warning("skipping corrupt pending event: %s", e)
        return events

    def clear(self) -> None:
        if self._path.exists():
            self._path.unlink()

    def replace(self, events: List[Event]) -> None:
        if not events:
            self.clear()
            return
        tmp = self._path.with_suffix(".jsonl.tmp")
        with open(tmp, "w") as f:
            for event in events:
                f.write(event.model_dump_json() + "\n")
        tmp.replace(self._path)


def _build_embedding_text(event: Event) -> str:
    if event.type == EventType.BROWSER:
        try:
            data = json.loads(event.raw_content)
            parts = [p for p in (data.get("title"), data.get("domain")) if p]
            text = " | ".join(parts) if parts else event.raw_content
        except (json.JSONDecodeError, TypeError):
            text = event.raw_content

    elif event.type == EventType.TERMINAL:
        try:
            data = json.loads(event.raw_content)
            text = data.get("command", event.raw_content)
        except (json.JSONDecodeError, TypeError):
            text = event.raw_content

    elif event.type in (EventType.GIT_COMMIT, EventType.GIT_STASH):
        try:
            data = json.loads(event.raw_content)
            parts = []
            if data.get("message"):
                parts.append(data["message"])
            files = data.get("files_changed", [])
            if files:
                parts.append("files: " + ", ".join(files[:20]))
            if data.get("diff"):
                parts.append(data["diff"])
            text = " | ".join(parts) if parts else event.raw_content
        except (json.JSONDecodeError, TypeError):
            text = event.raw_content

    else:
        text = event.raw_content

    if event.project:
        text = f"{text} [{event.project}]"

    return text.strip()


class EventPreprocessor:
    def __init__(self):
        self._pending_store = PendingEventStore()

    def generate_embedding(self, text: str) -> Optional[List[float]]:
        log.debug("generating embedding (len=%d chars)", len(text))
        try:
            vector = embedding.embed_one(text)
            log.debug("embedding generated: dim=%d", len(vector))
            return vector
        except Exception as e:
            log.error("embedding failed: %s: %s", type(e).__name__, e)
            raise

    def process_event(self, event: Event) -> List[Event]:
        log.debug("processing event type=%s source=%s", event.type, event.source)
        text = _build_embedding_text(event)
        if not text:
            log.debug("skipping event with empty text type=%s", event.type)
            return []
        chunks = chunk_text(
            text, embedding.effective_max_tokens(), config.embedding_chunk_overlap
        )

        if len(chunks) == 1:
            event.embedding = self.generate_embedding(chunks[0])
            return [event]

        log.debug("event split into %d chunks (type=%s)", len(chunks), event.type)
        results = []
        for i, chunk in enumerate(chunks):
            chunk_event = event.model_copy(deep=True)
            chunk_event.metadata = {
                **event.metadata,
                "chunk_index": i,
                "total_chunks": len(chunks),
            }
            if event.source_doc_id:
                chunk_event.source_doc_id = f"{event.source_doc_id}::chunk::{i}"
            chunk_event.embedding = self.generate_embedding(chunk)
            results.append(chunk_event)

        return results

    def process_events(self, events: List[Event]) -> List[Event]:
        processed = []
        for event in events:
            try:
                processed.extend(self.process_event(event))
            except Exception as e:
                log.warning("embedding failed, queuing for later: %s", e)
                self._pending_store.save(event)
        return processed

    def process_pending(self) -> List[Event]:
        pending = self._pending_store.load_all()
        if not pending:
            return []

        # Drain in bounded batches so a large backlog can't block the daemon
        # cycle or spike memory by embedding everything at once.
        limit = config.max_pending_per_cycle
        batch = pending[:limit]
        remainder = pending[limit:]

        # Expand events into (chunk_event, chunk_text) pairs
        pairs: List[tuple] = []
        for event in batch:
            text = _build_embedding_text(event)
            chunks = chunk_text(
                text, embedding.effective_max_tokens(), config.embedding_chunk_overlap
            )
            if len(chunks) == 1:
                pairs.append((event, chunks[0]))
            else:
                for i, chunk in enumerate(chunks):
                    chunk_event = event.model_copy(deep=True)
                    chunk_event.metadata = {
                        **event.metadata,
                        "chunk_index": i,
                        "total_chunks": len(chunks),
                    }
                    if event.source_doc_id:
                        chunk_event.source_doc_id = f"{event.source_doc_id}::chunk::{i}"
                    pairs.append((chunk_event, chunk))

        pairs = [(event, text) for event, text in pairs if text]
        if not pairs:
            # Nothing embeddable in this batch — drop it so it can't wedge the queue.
            self._pending_store.replace(remainder)
            return []

        texts = [text for _, text in pairs]
        try:
            vectors = embedding.embed_many(texts)
            result = []
            for (chunk_event, _), vector in zip(pairs, vectors):
                chunk_event.embedding = vector
                result.append(chunk_event)
            # Persist the unprocessed remainder; this batch is now embedded and
            # handed back to the caller to store.
            self._pending_store.replace(remainder)
            log.info(
                "bulk embedded %d/%d pending events (%d chunks, %d remaining)",
                len(batch),
                len(pending),
                len(result),
                len(remainder),
            )
            return result
        except Exception as e:
            # Leave the file untouched so the batch is retried next cycle.
            log.error("bulk embedding of pending events failed: %s", e)
            return []

    def clear_pending(self) -> None:
        self._pending_store.clear()

File: rust/patcha/src/process.rs
use crate::{
    config::Config,
    embedding::{chunk_text, Embedder},
    models::{Event, EventType},
};
use anyhow::Result;
use std::path::{Path, PathBuf};
use std::sync::Arc;

// ---------------------------------------------------------------------------
// PendingEventStore
// ---------------------------------------------------------------------------

/// JSONL file of events whose embedding failed — retried on next daemon cycle.
pub struct PendingEventStore {
    path: PathBuf,
}

impl PendingEventStore {
    pub fn new(data_dir: &Path) -> Self {
        Self {
            path: data_dir.join("pending_events.jsonl"),
        }
    }

    pub fn save(&self, event: &Event) -> Result<()> {
        if let Some(parent) = self.path.parent() {
            std::fs::create_dir_all(parent)?;
        }
        use std::io::Write;
        let mut f = std::fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open(&self.path)?;
        writeln!(f, "{}", serde_json::to_string(event)?)?;
        Ok(())
    }

    pub fn load_all(&self) -> Result<Vec<Event>> {
        if !self.path.exists() {
            return Ok(Vec::new());
        }
        let content = std::fs::read_to_string(&self.path)?;
        let mut events = Vec::new();
        for line in content.lines() {
            let line = line.trim();
            if line.is_empty() {
                continue;
            }
            match serde_json::from_str::<Event>(line) {
                Ok(e) => events.push(e),
                Err(e) => tracing::warn!(error=%e, "skipping corrupt pending event"),
            }
        }
        Ok(events)
    }

    pub fn clear(&self) -> Result<()> {
        if self.path.exists() {
            std::fs::remove_file(&self.path)?;
        }
        Ok(())
    }

    /// Atomically replace the file with `events` (or delete if empty).
    pub fn replace(&self, events: &[Event]) -> Result<()> {
        if events.is_empty() {
            return self.clear();
        }
        let tmp = self.path.with_extension("jsonl.tmp");
        {
            use std::io::Write;
            let mut f = std::fs::File::create(&tmp)?;
            for event in events {
                writeln!(f, "{}", serde_json::to_string(event)?)?;
            }
        }
        std::fs::rename(tmp, &self.path)?;
        Ok(())
    }
}

// ---------------------------------------------------------------------------
// Semantic text extraction
// ---------------------------------------------------------------------------

/// Build the text that will be embedded for a given event.
/// Mirrors Python's `_build_embedding_text`.
pub fn build_embedding_text(event: &Event) -> String {
    let text = match event.event_type {
        EventType::Browser => {
            if let Ok(data) = serde_json::from_str::<serde_json::Value>(&event.raw_content) {
                let parts: Vec<&str> = [
                    data.get("title").and_then(|v| v.as_str()),
                    data.get("domain").and_then(|v| v.as_str()),
                ]
                .into_iter()
                .flatten()
                .collect();
                if parts.is_empty() {
                    event.raw_content.clone()
                } else {
                    parts.join(" | ")
                }
            } else {
                event.raw_content.clone()
            }
        }

        EventType::Terminal => {
            if let Ok(data) = serde_json::from_str::<serde_json::Value>(&event.raw_content) {
                data.get("command")
                    .and_then(|v| v.as_str())
                    .unwrap_or(&event.raw_content)
                    .to_owned()
            } else {
                event.raw_content.clone()
            }
        }

        EventType::GitCommit | EventType::GitStash => {
            if let Ok(data) = serde_json::from_str::<serde_json::Value>(&event.raw_content) {
                let mut parts: Vec<String> = Vec::new();
                if let Some(msg) = data.get("message").and_then(|v| v.as_str()) {
                    parts.push(msg.to_owned());
                }
                if let Some(files) = data.get("files_changed").and_then(|v| v.as_array()) {
                    let file_list: Vec<&str> =
                        files.iter().filter_map(|v| v.as_str()).take(20).collect();
                    if !file_list.is_empty() {
                        parts.push(format!("files: {}", file_list.join(", ")));
                    }
                }
                if let Some(diff) = data.get("diff").and_then(|v| v.as_str()) {
                    parts.push(diff.to_owned());
                }
                if parts.is_empty() {
                    event.raw_content.clone()
                } else {
                    parts.join(" | ")
                }
            } else {
                event.raw_content.clone()
            }
        }

        EventType::Screen => {
            // Prepend the VLM gist so retrieval works on what the user was doing,
            // not just literal on-screen text: gist | app — window: ocr_text.
            match event.metadata.get("gist").and_then(|v| v.as_str()) {
                Some(gist) if !gist.is_empty() => format!("{gist} | {}", event.raw_content),
                _ => event.raw_content.clone(),
            }
        }

        _ => event.raw_content.clone(),
    };

    let text = text.trim().to_owned();
    if let Some(proj) = &event.project {
        if !proj.is_empty() {
            return format!("{text} [{proj}]");
        }
    }
    text
}

// ---------------------------------------------------------------------------
// EventPreprocessor
// ---------------------------------------------------------------------------

pub struct EventPreprocessor {
    embedder: Arc<Embedder>,
    pending_store: PendingEventStore,
    max_pending_per_cycle: usize,
    chunk_overlap: usize,
}

impl EventPreprocessor {
    pub fn new(cfg: &Config, embedder: Arc<Embedder>) -> Self {
        Self {
            embedder,
            pending_store: PendingEventStore::new(&cfg.data_dir),
            max_pending_per_cycle: cfg.max_pending_per_cycle,
            chunk_overlap: cfg.embedding_chunk_overlap,
        }
    }

    // -----------------------------------------------------------------------
    // Process a batch of fresh events
    // -----------------------------------------------------------------------

    /// Embed `events`, returning the embedded events.
    /// Events that fail to embed are saved to the pending store for retry.
    pub fn process_events(&self, events: Vec<Event>) -> Vec<Event> {
        let mut processed = Vec::new();
        for event in events {
            match self.process_one(event) {
                Ok(chunks) => processed.extend(chunks),
                Err(e) => {
                    tracing::warn!(error=%e, "embedding failed — queuing for retry");
                }
            }
        }
        processed
    }

    fn process_one(&self, event: Event) -> Result<Vec<Event>> {
        let text = build_embedding_text(&event);
        if text.is_empty() {
            return Ok(Vec::new());
        }

        let max_tokens = self.embedder.effective_max_tokens;
        let chunks = chunk_text(&text, max_tokens, self.chunk_overlap)?;

        if chunks.len() == 1 {
            let mut e = event;
            e.embedding = Some(self.embedder.embed_one(&chunks[0])?);
            return Ok(vec![e]);
        }

        tracing::debug!(
            chunks = chunks.len(),
            event_type = %event.event_type,
            "event split into multiple chunks"
        );

        let total = chunks.len();
        let mut results = Vec::with_capacity(total);
        for (i, chunk) in chunks.iter().enumerate() {
            let mut chunk_event = event.clone();
            chunk_event
                .metadata
                .insert("chunk_index".into(), serde_json::json!(i));
            chunk_event
                .metadata
                .insert("total_chunks".into(), serde_json::json!(total));
            if let Some(ref doc_id) = event.source_doc_id {
                chunk_event.source_doc_id = Some(format!("{doc_id}::chunk::{i}"));
            }
            chunk_event.embedding = Some(self.embedder.embed_one(chunk)?);
            results.push(chunk_event);
        }

        Ok(results)
    }

    // -----------------------------------------------------------------------
    // Retry pending events
    // -----------------------------------------------------------------------

    /// Re-embed up to `max_pending_per_cycle` events from the pending queue.
    /// Successfully embedded events are returned; the rest stay on disk.
    pub fn process_pending(&self) -> Result<Vec<Event>> {
        let pending = self.pending_store.load_all()?;
        if pending.is_empty() {
            return Ok(Vec::new());
        }

        let limit = self.max_pending_per_cycle;
        let batch = &pending[..limit.min(pending.len())];
        let remainder = &pending[limit.min(pending.len())..];

        // Build (event, chunk_text) pairs
        let mut pairs: Vec<(Event, String)> = Vec::new();
        let max_tokens = self.embedder.effective_max_tokens;

        for event in batch {
            let text = build_embedding_text(event);
            if text.is_empty() {
                continue;
            }
            let chunks = chunk_text(&text, max_tokens, self.chunk_overlap)?;
            let total = chunks.len();
            for (i, chunk) in chunks.into_iter().enumerate() {
                let mut chunk_event = event.clone();
                if total > 1 {
                    chunk_event
                        .metadata
                        .insert("chunk_index".into(), serde_json::json!(i));
                    chunk_event
                        .metadata
                        .insert("total_chunks".into(), serde_json::json!(total));
                    if let Some(ref doc_id) = event.source_doc_id {
                        chunk_event.source_doc_id = Some(format!("{doc_id}::chunk::{i}"));
                    }
                }
                pairs.push((chunk_event, chunk));
            }
        }

        if pairs.is_empty() {
            // Nothing embeddable — drop the batch so it can't wedge the queue.
            self.pending_store.replace(remainder)?;
            return Ok(Vec::new());
        }

        let texts: Vec<String> = pairs.iter().map(|(_, t)| t.clone()).collect();

        match self.embedder.embed_many(texts) {
            Ok(vectors) => {
                let results: Vec<Event> = pairs
                    .into_iter()
                    .zip(vectors)
                    .map(|((mut event, _), vec)| {
                        event.embedding = Some(vec);
                        event
                    })
                    .collect();

                self.pending_store.replace(remainder)?;
                tracing::info!(
                    embedded = results.len(),
                    batch_size = batch.len(),
                    remaining = remainder.len(),
                    "bulk embedded pending events"
                );
                Ok(results)
            }
            Err(e) => {
                // Leave the file untouched — retry next cycle.
                tracing::error!(error=%e, "bulk embedding of pending events failed");
                Ok(Vec::new())
            }
        }
    }

    pub fn save_pending(&self, event: &Event) -> Result<()> {
        self.pending_store.save(event)
    }

    pub fn clear_pending(&self) -> Result<()> {
        self.pending_store.clear()
    }
}
